x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java

```diff
@@ -127,6 +136,45 @@ private String getEnrichIndexKey(SearchRequest searchRequest) {
         return ia.getIndices().get(0).getName();
     }
 
+    private List<Map<?, ?>> toCacheValue(SearchResponse response) {
+        List<Map<?, ?>> result = new ArrayList<>(response.getHits().getHits().length);
+        for (SearchHit hit : response.getHits()) {
+            result.add(deepCopy(hit.getSourceAsMap(), true));
+        }
+        return Collections.unmodifiableList(result);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T deepCopy(T value, boolean unmodifiable) {
+        return (T) innerDeepCopy(value, unmodifiable);
+    }
+
+    private static Object innerDeepCopy(Object value, boolean unmodifiable) {
```

Comment:
I'm not as familiar with all the data types that may be present in the map from a `SearchHit`, but `IngestDocument::deepCopy` handles a couple types that this method does not. ([link](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestDocument.java#L742))

Reply:
Initially I looked at the `IngestDocument#deepCopy(...)`  method, but then realised that some of the data type (e.g. `Set`, `long[]`, `Date`) are never in the search hit. The `SearchHit.getSourceAsMap()` method uses `XContentHelper#convertToMap(...)` method, which eventually invokes `AbstractXContentParser#readValueUnsafe(...)` via `XContentParser#map()` method and I think the mentioned types are never produced by the `readValueUnsafe()` method.

I think we can reuse `IngestDocument#deepCopy(...)`. There is some code duplication between the two methods. But this method will then need to be modified to optionally wrap list / map in an unmodifiable wrapper and this would only be used by `EnrichCache`. I currently lean towards keeping this method, since the code duplication is minimal and it is targeted for creating a deep copy of search hits. But if we keep this method then we need a good unit test for it as well. What do you think?